### DISCUSSION TITLE
Support for fractional nuclear charges

### DISCUSSION THREAD

**Discussion by JFurness1** [pyscf/pyscf#2708] [Q&A]
Dear Developers,

I am looking to use PySCF to perform calculations on imaginary systems of nuclei with fractional nuclear charges, along the lines of our recent pre-print at [arXiv:2103.03178](https://arxiv.org/abs/2103.03178). Specifically, I would like to use the high-accuracy low dimensional periodic boundary condition methods available.

Looking through the manual and dipping into the code it appears that non-integer nuclear charges are not currently supported. Is this correct?

If they are not supported I am happy to implement such functionality myself and contribute the changes to the project, if desired by the maintainers. It seems to my naive assessment that basic functionality should be a simple enough case of reading inputs of fractional charges and passing them into the nuclear-electron attraction and nuclear-nuclear repulsion routines. I expect properly supporting symmetry may require more work.

If support for fractional nuclear charges is desired in the main project, is there a preferred forum for getting advice on how best to implement this in line with the current structures?

Many thanks,
James

**Comment by JFurness1** [comment]
Hmm... Starting out on exploring this it seems like the expectation for integer nuclear charges is pretty deeply baked into libcint, which is disappointing.

I guess the solution I will follow will be to hook into the python that calls the nuclear-electron integrals and call my own integral routine as well. Not a very clean solution though.

**Comment by gkc1000** [comment]
You should not need to go to the libcint level; the electron-nuclear integral with fractional charges can be obtained by scaling the integrals from integer charges. However, in a periodic system, the system will not be neutral unless you have a fractional number of electrons also. Without assumption of neutrality, the total energy will formally be infinite.

**Comment by JFurness1** [comment]
Thanks! I realised this last night as I was falling asleep that a better route would be to calculate the nuclear integrals for each atom separately, then scale them appropriately.

Thanks for the warning about the periodic systems. We have been constraining ourselves to charge neutral species to avoid the complications of unbound electrons, and things like this clearly make that constraint critical.

Looks like I have everything I need to implement it. I'll try to do so in as clean a way as possible, but I expect this may be outside the scope of what PySCF wants in the main code base.

Thanks for your comments.

**Comment by sunqm** [comment]
Here is a relevant snippet

https://github.com/pyscf/pyscf/blob/master/pyscf/qmmm/itrf.py#L134

On Fri, Mar 26, 2021 at 08:57:01AM -0700, James Furness wrote:
> Thanks! I realised this last night as I was falling asleep that a better route would be to calculate the nuclear integrals for each atom separately, then scale them appropriately.
> 
> Thanks for the warning about the periodic systems. We have been constraining ourselves to charge neutral species to avoid the complications of unbound electrons, and things like this clearly make that constraint critical.
> 
> Looks like I have everything I need to implement it. I'll try to do so in as clean a way as possible, but I expect this may be outside the scope of what PySCF wants in the main code base.
> 
> Thanks for your comments.
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/pyscf/pyscf/issues/862#issuecomment-808329680

--

**Comment by JFurness1** [comment]
Thank you for the ongoing support with this. 

I have successfully implemented fractional nuclear charges for molecular systems and have confirmed that they agree with my reference to high precision.

I am now looking at generalising to periodic calculations and I am having trouble getting 1D DFT calculations to give sensible results even for the simple integer H2 chain case. I expect I have a fundamental misunderstanding somewhere, but after reading through the relevant sections of documentation I am still confused.

I am taking a system of H-H around equilibrium (~1.4 Bohr) and expecting that when the unit cell is made very large (20 Bohr) this will give roughly the same energy as the isolated H-H system.

For this 1D test chain I am running

```python
from pyscf.pbc import scf, df, gto

cell = gto.Cell()
cell.exp_to_discard = 0.1
cell.build(
    unit='B',
    a = [[20.0,0,0],[0,1,0],[0,0,1]],
    atom='''H@1 0 0 0; H@2 1.4 0 0.0''',
    basis = 'cc-pVTZ',
    dimension = 1,
    verbose = 4)
kpts = cell.make_kpts([6,1,1])
mf = scf.KUKS(cell, kpts)
mf.with_df = df.MDF(cell)

print(mf.kernel())
```
reporting
`converged SCF energy = -0.647489336562256`

Running the isolated H-H (R = 1.4 Bohr) calculation
```python
import numpy as np
from pyscf import scf, gto

mol = gto.Mole()
mol.build(
    unit='B',
    atom='''H@1 0 0 0; H@2 1.4 0 0.0''',
    basis = 'cc-pVTZ',
    verbose = 4)
mf = scf.UKS(mol)

print(mf.kernel())
```
reports
`converged SCF energy = -1.13685518798485  <S^2> = 4.4408921e-15  2S+1 = 1`

My experiments with other codes suggest that the highly separeted 1D chain should match the molecular calculation closely, assuming the output is total energy per unit cell in Hartree.

I would welcome any advice on where I might be going wrong here.

Many thanks,
James

**Comment by gkc1000** [comment]
Box needs to be big in all dimensions.

**Comment by JFurness1** [comment]
Thanks, that makes some sense. I think I had assumed that in low dimensions the non-periodic directions would be independent of a.

However, I have been experimenting a bit and it seems I am still missing something. If I increase the size of the box and add the `mesh` option to meet the warnings,
`WARN: mesh [83 39 39] is not enough for AFTDF.get_nuc function to get integral accuracy 1e-08.
Recommended mesh is [ 84 167 167].`

Then my script becomes this:

```python
from pyscf.pbc import scf, df, gto

cell = gto.Cell()
cell.exp_to_discard = 0.1
cell.build(
    unit='B',
    a = [[20.0,0,0],[0, 40,0],[0,0,40]],
    mesh = [ 84, 167, 167],
    atom='''H@1 0 0 0; H@2 1.4 0 0.0''',
    basis = 'cc-pVTZ',
    dimension = 1,
    verbose = 4)
kpts = cell.make_kpts([6,1,1])
mf = scf.KUKS(cell, kpts)
mf.with_df = df.MDF(cell)

print(mf.kernel())
```

which returns

`converged SCF energy = -0.756347795838647`

Different than before (and with much increased computation time), but still far from the expected ~ -1.14.

**Comment by gkc1000** [comment]
Actually I had not noticed the dimension = 1 in the input. For sanity, can
you make a minimal basis (so no small exponents) and use GDF and report the
result.

**Comment by JFurness1** [comment]
Certainly. I have checked out the master branch of the main repo before running this, but the same result is found on my local modified branch. 

If I use STO-3G then the smallest exponent is 0.16885540.

A minimal example using GDF is then

```python
from pyscf.pbc import scf, df, gto

cell = gto.Cell()
cell.build(
    unit='B',
    a = [[20.0,0,0],[0, 20,0],[0,0,20]],
    atom='''H@1 0 0 0; H@2 1.4 0 0.0''',
    basis = 'sto-3g',
    dimension = 1,
    verbose = 4)

cell.build()
kpts = cell.make_kpts([16,1,1])
mf = scf.KUKS(cell, kpts)
mf.with_df = df.GDF(cell)

foo = mf.kernel()
```

The simple output is 
`converged SCF energy = -0.774156207589776`
showing the problem.

The full output is attached 
[minimal.txt](https://github.com/pyscf/pyscf/files/6232274/minimal.txt)

**Comment by gkc1000** [comment]
Odd. @fishjojo can you take a look? Also try with dimension = 3.

On Tue, Mar 30, 2021 at 2:21 PM James Furness ***@***.***>
wrote:

> Certainly. I have checked out the master branch of the main repo before
> running this, but the same result is found on my local modified branch.
>
> If I use STO-3G then the smallest exponent is 0.16885540.
>
> A minimal example using GDF is then
>
> from pyscf.pbc import scf, df, gto
> cell = gto.Cell()cell.build(
>     unit='B',
>     a = [[20.0,0,0],[0, 20,0],[0,0,20]],
>     ***@***.*** 0 0 0; ***@***.*** 1.4 0 0.0''',
>     basis = 'sto-3g',
>     dimension = 1,
>     verbose = 4)
> cell.build()kpts = cell.make_kpts([16,1,1])mf = scf.KUKS(cell, kpts)mf.with_df = df.GDF(cell)
> foo = mf.kernel()
>
> The simple output is
> converged SCF energy = -0.774156207589776
> showing the problem.
>
> The full output is attached
> minimal.txt <https://github.com/pyscf/pyscf/files/6232274/minimal.txt>
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/pyscf/pyscf/issues/862#issuecomment-810585785>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAN5NRCYW3QWN6ZEXFKMZS3TGI6D7ANCNFSM4ZZX6BNQ>
> .
>
-- 
Sent from Gmail Mobile

**Comment by JFurness1** [comment]
`dimension = 3` works as expected, I've been able to continue working on my project using this as a work around, but obviously we should understand what is going wrong with the low dimensional case.

**Comment by fishjojo** [comment]
For DFT with density fitting, you'll need to call the function as follows
```
mf = scf.KUKS(cell, kpts).density_fit()
```
This will use Becke grids for xc functional evaluation.
The way that you invoke density fitting will use the uniform grid to evaluate xc functionals. This works in 3D in your case because there are enough grid points in all three directions. But for the low dimension case, the uniform grid is much smaller. You can print out `cell.mesh` to check.

**Comment by JFurness1** [comment]
Thank you, this advice has been useful.

I have had to deal with other tasks for a week or so, but I am returning to this now.

While the molecular fractional charges seem to be working as intended, I'm struggling a little to generalise to periodic systems, probably due to my unfamiliarity with the methods.

I have been exploring a simple KRKS calculation with a debugger

```python
za = 0.9
zb = 1.1
r = 6.0
cell = gto.Cell()
cell.exp_to_discard = 0.1
cell.build(
    unit='B',
    a = [[2*r,0,0],[0, 10,0],[0,0,10]],
    atom="H@1 0 0 0; H@2 {:.1f} 0 0.0".format(r),
    basis = 'cc-pVDZ',
    verbose = 4,
    nuclear_charges={'H@1':za, 'H@2':zb},
    dimension = 3)

cell.build()
kpts = cell.make_kpts([4,1,1])

mf = scf.KRHF(cell)
mf.kpts = kpts
mf = scf.addons.smearing_(mf, sigma=.1, method='fermi')

mf = mf.mix_density_fit()
ehf = mf.kernel()
```

Stepping through, I find that the core Hamiltonian building `pyscf/pbc/scf/khf.py` `get_hcore()` calls `pyscf/pbc/df/aft.py` `get_nuc()`, which in turn calls `get_pp_loc_part1` which seems to be the function I need to modify. 

Looking inside, I find it calls `charges = cell.atom_charges()`, which I have already modified to return the fractional charges.

```python
charges = cell.atom_charges()
print(charges)
# prints [0.9 1.1] as expected.
```
`charges` appears to be used in some `einsum` calls, so there should be no problem with the change to floats.

It seems like the same change to `atom_charges()` covers the Ewald sum for the nuclear repulsion as well.

Naively, I thought this would be sufficient, but now I find a problem. The total energy now decreases roughly linearly with increasing number of K points, never reaching a convergence. When nuclear charges are left as 1.0 and 1.0 convergence is reached quickly. So something is wrong.

Reading a little further, I guess what I have done is modified the long range portion of the electron-nuclei Ewald sum, but not the short range. 

Digging in further I find that `_compensate_nuccell()` prepares data for "compensated Gaussian charges", and copyies `cell._atm`, which contains the unmodified integer charges usually associated with the elements. 

Then `_int_nuc_vloc()` then appends a copy of `._atm` with parameters for a tight Gaussian function, copying the integer charges again.

Then  `pyscf/pbc/df/incore.py` `aux_e2()` is called which proceeds to send the data into `libcint` three center integral routines.

```python
    int3c = wrap_int3c(cell, auxcell, intor, aosym, comp, kptij_lst, **kwargs)
    out = numpy.empty((nkptij,comp,nao_pair,naux), dtype=dtype)
    out = int3c(shls_slice, out)
```
Looking into the dimensions of `out` seems to imply `(out = (nkpt, 1, basis functions?, 2*natom)`

Naively, it seems to me that I can re-scale `out` to the fractional charges. I attempted to do this with,

```python
    if cell.nuclear_charges:
        charges = cell.atom_charges()
        nc = len(charges)
        for i, c in enumerate(charges):
            scale = c/cell._atm[i,0]
            out[:, :, :, i] *= scale         # _atm entry from cell
            out[:, :, :, i + nc] *= scale    # second _atm entry from tight Gaussian
```

But the same problem of linearly decreasing energy with increasing number of k points still remains.

At this point I feel like I am reaching around in the dark. If anyone can suggest any aspects I might have missed, or point me in the right direction, that would be greatly appreciated. 

Additionally, if there are any recommended resources for understanding the practical implementation of the nuclear attraction integrals I would welcome suggestions. I am currently working through [Sun 2020](https://doi.org/10.1063/5.0006074) and [Sun 2017](dx.doi.org/10.1063/1.4998644) to try to develop a better understanding.

Many thanks,
James

**Reply by sunqm** [reply]
For a quick solution, you can modify this class, by assigning a custom cell to this class, as shown in this code 
https://github.com/pyscf/pyscf/blob/36d59679a18429756199bfb838fd141e3d4161dd/pyscf/pbc/df/rsdf_builder.py#L1103
The custom cell holds the coordinates of the point charges. You can modify this function to create custom Cell object https://github.com/pyscf/pyscf/blob/36d59679a18429756199bfb838fd141e3d4161dd/pyscf/qmmm/mm_mole.py#L125

Then changing the associated charges
https://github.com/pyscf/pyscf/blob/36d59679a18429756199bfb838fd141e3d4161dd/pyscf/pbc/df/rsdf_builder.py#L1175

calling the get_nuc method would produce a potential from fractional charges. However, this potential excludes the contribution from G=0.

**Comment by sevkel** [comment]
Hey, im a undergrad student and new to PySCF and saw this issue here while i was trying to plot fractional charge behaviour for the H atom with different DFT functionals. But i really struggle to do so, can anyone here give me a hint. It seems not to easy to implement. This is my code so far. i thought the mm_charge method could deal with it but i dont really understand it.

mol = gto.M(atom='H 0 0 0', basis='ccpvdz', verbose=0)
mol.spin = 1
mol.build()

mftemp = dft.UKS(mol, xc='pbe')

energies = []
elecs = np.arange(9, 10, 0.1)
for charge in elecs:
    mf = itrf.mm_charge(mftemp, [(0.6,0.6,0.6)], [elecs])
    mf.kernel()
    energies.append(mf.e_tot)

Thanks in advance!
